File: 프로그래머스/4/64063. 호텔 방 배정/호텔 방 배정.cpp
#include <string>
#include <vector>
#include <unordered_map>

using namespace std;

unordered_map<long long, long long> m;

long long find(long long idx) {
    if(m.find(idx) == m.end()) {
        m[idx] = idx + 1;
        return idx;
    }
    else {
        long long nxtIdx = find(m[idx]);
        m[idx] = nxtIdx + 1;
        return nxtIdx;
    }
}

vector<long long> solution(long long k, vector<long long> room_number) {
    vector<long long> answer;
    
    for(long long rn: room_number) {
        long long idx = find(rn);
        answer.push_back(idx);
    }
    
    return answer;
}

File: 프로그래머스/4/64063. 호텔 방 배정/호텔 방 배정.py
import sys

sys.setrecursionlimit(1000000)

def find_empty(idx):
    global goto
    
    if idx not in goto:
        goto[idx] = idx + 1
        return idx
    
    ret = find_empty(goto[idx])
    goto[idx] = ret + 1
    return ret

def solution(k, room_number):
    global goto
    
    answer = []
    goto = {}
    
    for rn in room_number:
        answer.append(find_empty(rn))
    
    return answer